### PR TITLE
BAAS-21844: estimate mem usage for large objects

### PR DESCRIPTION
**Pull request by Gabri3l**
I tested this in `baas` with a function like

```
exports = async function(arg){
  i = 0;
  testObj = {};
  while (true) {
    testObj[i] = "a";
    i++;
  }
  return "hello";
};
```

the before and after was quite a big difference, with this change we threw an error much faster!

This might not address nested objects, but the sampling will mitigate that. We would probably need a follow up to this change to better take that into account.

### CODE REVIEW

File: memory_test.go [hunk 1]
```diff
@@ -517,7 +524,7 @@ func TestMemArraysWithLenThreshold(t *testing.T) {
 			2 + SizeEmpty + SizeNumber + (50 * SizeNumber) + (50 * SizeBool),
 		},
 		{
-			"mixed array under threshold but over limit",
+			"array under threshold but over limit",
```
Comment (Gabri3l) [inline memory_test.go:527]: Just  drive-by

**Review by LijieZhang1998 — APPROVED** [review]
LGTM 🥇

File: object.go [hunk 2]
```diff
@@ -1925,6 +1925,37 @@ func (i *privateId) string() unistring.String {
...
+		v := o.values[k]
+		if v == nil {
+			continue
+		}
+
+		inc, err := v.MemUsage(ctx)
+		samplesVisited += 1
+		total += inc
+		kMem := uint64(len(k))
+		total += kMem
```
[the remark this answers is not in the document]
Comment (Gabri3l) [reply object.go:1949]: omg sorry that was a change I had to log it when debugging something ahah, thanks for pointing it out!

File: object.go [hunk 3]
```diff
@@ -1936,20 +1967,27 @@ func (o *baseObject) MemUsage(ctx *MemUsageContext) (uint64, error) {
...
+	if ctx.ObjectPropsLenExceedsThreshold(len(o.propNames)) {
+		inc, err := o.estimateMemUsage(ctx)
 		total += inc
-		total += uint64(len(k))
-
 		if err != nil {
 			return total, err
 		}
+	} else {
+		for _, k := range o.propNames {
```
[the remark this answers is not in the document]
Comment (Gabri3l) [reply object.go:1977]: I think 1k is definitely high, especially with objects. I'm down to set it to 100 or so, the baas PR will include the value

File: object.go [hunk 4]
```diff
@@ -1925,6 +1925,37 @@ func (i *privateId) string() unistring.String {
 	return privateIdString(i.name)
 }
 
+// estimateMemUsage helps calculating mem usage for large objects.
+// It will sample the object and use those samples to estimate the
+// mem usage.
+func (o *baseObject) estimateMemUsage(ctx *MemUsageContext) (uint64, error) {
```
[the remark this answers is not in the document]
Comment (Gabri3l) [reply object.go:1931]: I'll file tickets for them, I haven't seen OOMs around those objects but at this point it should be a pretty easy change so it doesn't hurt to get ahead.

[the remark this answers is not in the document]
Comment (Gabri3l) [reply object.go:1931]: Alright created https://jira.mongodb.org/browse/BAAS-22177 and https://jira.mongodb.org/browse/BAAS-22176
